core/cloudflow-spark/src/main/scala/cloudflow/spark/kafka/SparkStreamletContextImpl.scala

```diff
@@ -58,12 +54,7 @@ class SparkStreamletContextImpl(
 
     val rawDataset = src.select($"value").as[Array[Byte]]
 
-    val dataframe: Dataset[Row] = rawDataset.mapPartitions { iter ⇒
-      val avroDecoder = new SparkAvroDecoder[In](schema)
-      iter.map(avroDecoder.decode)
-    }(inRowEncoder)
-
-    dataframe.as[In]
+    rawDataset.map(inPort.codec.decode(_))
```

Comment:
@blublinsky Don't you get a `SerializationError` if you do this? Did you try on a cluster?
Back then, I had many issues with `avro.Schema` because it's not serializable. 
I would expect that applying a transformation where the codec instance is is passed will attempt to serialize the `Schema` member of it.

Reply:
It would be great if it works. It makes it follow the same pattern that we are using everywhere else.

Reply:
Yes it does. Just finished testing it on the cluster. The key here is that Avro generated class is just a simple case class, for which Spark creates a default encoder, that is used here.

Reply:
Update: bummer... 
I don't think it's going to work. 
I created a serialization test as follows:
```scala
class SerializationSpec extends WordSpec with SparkScalaTestSupport with TryValues {
  "Avro Codec" should {
    "be serializable" in {
      val avroInlet = AvroInlet[Data]("name")
      val deser     = avroInlet.codec.decode _
      try {
        SparkEnv.get.closureSerializer.newInstance().serialize(deser)
        // pass!
      } catch {
        case ex: Throwable => fail(ex.toString)
      }
    }
"Sanity check" should {
    "not be serializable" in {
      val localResource   = new Thread()
      val notSerializable = () => localResource
      try {
        SparkEnv.get.closureSerializer.newInstance().serialize(notSerializable)
        fail("Thread should not be serializable")
      } catch {
        case ex: Throwable => () // pass
      }

    }
  }
  }
```
and it fails with the same Exception that I was getting back in the day.
```
[info] - should be serializable *** FAILED ***
[info]   java.io.NotSerializableException: org.apache.avro.Schema$RecordSchema
[info]   Serialization stack:
[info]          - object not serializable (class: org.apache.avro.Schema$RecordSchema, value: {"type":"record","name":"Data","namespace":"cloudflow.spark.avro","fields":[{"name":"id","type":"int"},{"name":"name","type":"string"}]})
[info]          - field (class: cloudflow.streamlets.avro.AvroCodec, name: avroSchema, type: class org.apache.avro.Schema)
[info]          - object (class cloudflow.streamlets.avro.AvroCodec, cloudflow.streamlets.avro.AvroCodec@64f4c593)
[info]          - element of array (index: 0)
[info]          - array (class [Ljava.lang.Object;, size 1)
[info]          - field (class: java.lang.invoke.SerializedLambda, name: capturedArgs, type: class [Ljava.lang.Object;)
[info]          - object (class java.lang.invoke.SerializedLambda, SerializedLambda[capturingClass=class cloudflow.spark.SerializationSpec, functionalInterfaceMethod=scala/Function1.apply:(Ljava/lang/Object;)Ljava/lang/Object;, implementation=invokeStatic cloudflow/spark/SerializationSpec.$anonfun$new$3:(Lcloudflow/streamlets/avro/AvroCodec;[B)Lcloudflow/spark/avro/Data;, instantiatedMethodType=([B)Lcloudflow/spark/avro/Data;, numCaptured=1])
[info]          - writeReplace data (class: java.lang.invoke.SerializedLambda)
[info]          - object (class cloudflow.spark.SerializationSpec$$Lambda$12337/821591673, cloudflow.spark.SerializationSpec$$Lambda$12337/821591673@629795ce) (SerializationSpec.scala:33)
[info] Sanity check
[info] - should not be serializable
```

Reply:
Of course it will not. You do not have encoder. Take a look at my tests https://github.com/lightbend/Spark-sensors-proto. All works as a champ

Reply:
Verified that everything works on the cluster.